### PR TITLE
Update DeprecationMap to DynamicMap

### PR DESCRIPTION
**Pull request by jdconrad**
This renames DeprecationMap to DynamicMap, and changes the deprecation messages Map to accept a Map of String (keys) to Functions (updated values) instead. This creates more flexibility in either logging or updating values from params within a script. This change is required to fix (https://github.com/elastic/elasticsearch/issues/52103) in a future PR.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/script/AbstractSortScript.java [hunk 1]
```diff
@@ -29,14 +31,25 @@
...
 abstract class AbstractSortScript implements ScorerAware {
 
-    private static final Map<String, String> DEPRECATIONS = Map.of(
-            "doc",
-            "Accessing variable [doc] via [params.doc] from within a sort-script is deprecated in favor of directly accessing [doc].",
-            "_doc",
-            "Accessing variable [doc] via [params._doc] from within a sort-script is deprecated in favor of directly accessing [doc].");
+    private static final DeprecationLogger deprecationLogger =
+            new DeprecationLogger(LogManager.getLogger(DynamicMap.class));
+    private static final Map<String, Function<Object, Object>> FUNCTIONS = Map.of(
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/AbstractSortScript.java:40]: I might name this like `PARAMS_FUNCTIONS` or something like that, so it is clear they are part of `params`.

File: server/src/main/java/org/elasticsearch/script/DynamicMap.java [hunk 2]
```diff
@@ -19,28 +19,20 @@
...
 import java.util.Collection;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.Function;
 
-public final class DeprecationMap implements Map<String, Object> {
-
-    private static final DeprecationLogger deprecationLogger =
-        new DeprecationLogger(LogManager.getLogger(DeprecationMap.class));
+public final class DynamicMap implements Map<String, Object> {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/DynamicMap.java:27]: Could you please add a brief javadoc explaining what this class is?

**Review by rjernst — APPROVED** [review]
LGTM. Just a couple minor comments

File: server/src/main/java/org/elasticsearch/script/DynamicMap.java (hunk 2 above)
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/script/DynamicMap.java:27]: Added.

File: server/src/main/java/org/elasticsearch/script/AbstractSortScript.java (hunk 1 above)
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/script/AbstractSortScript.java:40]: Changed.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review. Will merge once CI finishes successfully.

**Comment by jdconrad** [comment]
@elastichmachine run elasticsearch-ci/bwc elasticsearch-ci/default-distro

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Pull request merged**